README.md
# Logstash Plugin

This gem is a plugin for [Logstash](https://github.com/elastic/logstash). During filter it takes one or more fields and uses that as input to query additional information. The original purpose is to enrich IP addresses with matching subnet, netname and hostname, but it is generic so that any field can be looked up. The function is similar to the translate filter's dictionary lookup, which supports files and regex. The jdbc_streaming filter plugin is also very useful if the data resides in a database. This plugins features are web based lookups and redis caching, for fast lookups.

## Documentation

weblookup {
    fields => ['[client][ip]']
    destinations => ['net']
    url => "http://localhost/ripe.php?ip=<item>"
    use_redis => true
    redis_path => "/var/run/redis/redis-server.sock"
    normalize => true
    newroot => "[records][properties]"
    roottostrip => "[records]"
}

Where <item> will be replaced by the value of client.ip

The first three components are needed for the plugin, the others are optional. use_redis and redis_path are for caching the response, this speedsup the requists. It's also possible to hardcode values here, but I'm not using it myself yet. normalize, newroot and roottostrip probably would be better in a separte plugin, but for now weblookup can move the json objects inside the roottostrip into it's own root, by default elasticsearch uses _source as invisible root.

## Need Help?

Need help? Raise an issue on https://github.com/janmg/logstash-filter-weblookup 



CHANGELOG.md
## 0.1.4
  - Attempt to continue gracefully if either source or destination is empty

## 0.1.3
  - Slightly better rescue handling when json parsing or replanting fails

## 0.1.2
  - Handle negative response better so that the response is always JSON, even if it's empty
 
## 0.1.1
  -

## 0.1.0
  - Plugin created with the logstash plugin generator


spec/filters/lookup_spec.rb
# encoding: utf-8
require_relative '../spec_helper'
require 'logstash/filters/lookup'
require 'redis'

class LogStash::Codecs::JSON
end

describe LogStash::Filters::Lookup do
    describe "test ip lookup" do
    let(:config) do <<-CONFIG
      filter {
        lookup {
          fields => ["ClientIP"]
          list => {
	     '127.0.0.1' => '{"ip":"127.0.0.1", "subnet":"127.0.0.0/8", "netname":"localnet", "hostname"="localhost"}'
	     '192.168.0.1' => '{"ip":"192.168.0.1", "subnet":"192.168.0.0/24", "netname":"private", "hostname":"router"}'
          }
        }
      }
    CONFIG
    end

    message = '{"ClientIP" : "23.100.57.65"}'
    sample("message" => message) do
        #insist { subject["clientIP"][0].to_s } == "192.168.0.1"
	insist { find(subject["ClientIP"]) } == '{"ip":"192.168.0.1", "subnet":"192.168.0.0/24", "netname":"private", "hostname":"router"}'
    end

    sample("message" => message) do
        #ip = subject[fields][0].to_s
	ip = '23.100.57.65'
        @red = Redis.new
        @red.del(ip)
	insist { find(ip) } == '{"ip":"23.100.57.65","netname":"Azure europenorth","subnet":"23.100.48.0\/20","hostname":"monitoring"}'
	@red.del(ip)
	@red.add(ip,'{"ip":"23.100.57.65","netname":"dummy"}')
        insist { find(ip) } == '{"ip":"23.100.57.65","netname":"dummy"}'
	@red.del(ip)
    end
  end
end


lib/logstash/filters/lookup.rb
# encoding: utf-8
require 'logstash/filters/base'
require 'redis'
require 'json'
require 'net/http'
require 'connection_pool'
require 'addressable/uri'

class LogStash::Filters::Lookup < LogStash::Filters::Base

  # This is how you configure this filter from your Logstash config.
  # [source,ruby]
  # ----------------------------------
  # filter {
  #    lookup {
  #       fields => ["ClientIP"]
  #       lookup => "http://127.0.0.1/ripe.php?ip=<item>&TOKEN=token"
  #       list => {
  #           '127.0.0.1' => '{"ip":"127.0.0.1", "subnet":"127.0.0.0/8", "netname":"localnet", "hostname":"localhost"}',
  #           '192.168.0.1' => '{"ip":"192.168.0.1", "subnet":"192.168.0.0/24", "netname":"private", "hostname":"router"}'
  #       }
  #       use_redis => false
  #    }
  # }
  # ----------------------------------
  #
  config_name "lookup"

  config :fields, :validate => :array, :required => true, :default => ["ClientIP"]
  # {"ip":"8.8.8.8","netname":"Google","subnet":"8.8.8.0\/24","hostname":"google-public-dns-a.google.com"}
  # In the query parameter has the <ip> tag will be replaced by the IP address to lookup, other parameters are optional and according to your lookup service. 
  config :destinations, :validate => :array, :required => false, :default => ["message"]

  # {"ip":"8.8.8.8","netname":"Google","subnet":"8.8.8.0\/24","hostname":"google-public-dns-a.google.com"}
  # In the query parameter has the <ip> tag will be replaced by the IP address to lookup, other parameters are optional and according to your lookup service. 
  config :lookup, :validate => :string, :required => false, :default => "http://localhost/ripe.php?ip=<item>&TOKEN=token"

  # Optional ruby hash with the key as a string and the value as a string in the form of a JSON object. These key's will not be looked up.
  config :list, :validate => :hash, :required => false

  # Optional Redis IP cache
  config :use_redis, :validate => :boolean, :required => false, :default => false
  config :redis_path, :validate => :string, :required => false
  config :redis_expiry, :validate => :number, :required => false, :default => 604800 
  config :normalize, :validate => :boolean, :required => false, :default => false

  HTTP_OPTIONS = {
      keep_alive_timeout: 300
  }

public
def register
    if use_redis
        unless redis_path.to_s.strip.empty?
            @red = Redis.new(path: redis_path)
        else
            @red = Redis.new()
        end
    end

    # input fields and destinations
    @is_one_destination=false
    if destinations.size == 1
        @logger.info("one destination found, it is #{destinations[0]}")
	@is_one_destination=true
    else
        if destinations.size != fields.size
            @logger.error("Configuration error, there must be an equal amount of destinations and fields, defaulting to using the field as a root for the new values. e.g. if the lookup is done on the value of [\"ClientIP\"] the destination will be [\"ClientIP\"][\"Key\"]")
            destinations=fields
        end
    end

    # http connectionpool
    @uri = Addressable::URI.parse(lookup)
    @uri.merge!(HTTP_OPTIONS)
    #@http = Net::HTTP.new(uri.host, uri.port, HTTP_OPTIONS)
    @uri.port=80 if (@uri.port.nil? && @uri.scheme=="http")
    @uri.port=443 if (@uri.port.nil? && @uri.scheme=="https")
    # find the key where the value is <item>, otherwise just use the value
    @params = @uri.query_values(Hash)
    @params.each do |key, value|
        if value="\<item\>" 
            @ip=key
	    @params.delete(key)
	    logger.info("the ip key in the uri is #{@ip}")
        end
    end
    @connpool = ConnectionPool.new(size: 4, timeout: 180) { 
        Net::HTTP.new(@uri.host, @uri.port)
    }
end # def register

def filter(event)
    if destinations[0] == "srcdst"
        # ... do special sauce
        src = parse(event.get(fields[0]).to_s)
	dst = parse(event.get(fields[1]).to_s)
        srcdst = { :srcnet => src["netname"], :srchost => src["hostname"], :dstnet => dst["netname"], :dsthost => dst["hostname"] }
        event.set("srcdst", srcdst)
        event.get("[srcdst]").each {|k, v| event.set(k, v) }
        event.remove("[srcdst]")
	@logger.trace("processed: #{event.get(fields[0]).to_s} #{src} #{event.get(fields[1]).to_s} #{dst} #{srcdst}")
    else
        fields.each_with_index do |field, index|
            # @logger.info(event.get("["+field+"]"))
            json = parse(event.get(field).to_s)
            event.set("["+destinations[index]+"]", json)
        end
    end
    # filter_matched should go in the last line of our successful code
    filter_matched(event)
end # def filter



private
def parse(field)
    x = find(field)
    begin
        json = JSON.parse(x)
    rescue JSON::ParserError
        json = x
    end			
end

def find(item)
    res = nil
    # Is item in list? (list is an optional hash)
    unless list.nil? 
        return list[item]
    end
    # Is item in redis?
    unless @red.nil?
        res = @red.get(item)
	unless res.nil?
            return res
        end
    end

    # find the key where the value is <item>, otherwise just use the value
    current_uri = @uri
    current_uri.query_values = @params.merge({@ip => item})
    #logger.info(@uri.to_s)
    @connpool.with do |conn|
        res = conn.request_get(current_uri).read_body
	#logger.info(res.to_s)
        unless @red.nil?
            @red.set(item, res)
            @red.expire(item,redis_expiry)
        end
    end
    return res
end

def normalize(event)
      event.set("net", JSON.parse(net))
      event.get("[records][properties]").each {|k, v| event.set(k, v) }
      event.remove("[records]")
      event.remove("[message]")
      return event
end

end # class LogStash::Filters::Lookup


lib/logstash/filters/weblookup.rb
# encoding: utf-8
require 'logstash/filters/base'
require 'redis'
require 'json'
require 'net/http'
require 'connection_pool'
require 'addressable/uri'

class LogStash::Filters::Webookup < LogStash::Filters::Base

  # This is how you configure this filter from your Logstash config.
  # [source,ruby]
  # ----------------------------------
  # filter {
  #    weblookup {
  #       fields => ["ClientIP"]
  #       url => "http://127.0.0.1/ripe.php?ip=<item>&TOKEN=token"
  #       list => {
  #           '127.0.0.1' => '{"ip":"127.0.0.1", "subnet":"127.0.0.0/8", "netname":"localnet", "hostname":"localhost"}',
  #           '192.168.0.1' => '{"ip":"192.168.0.1", "subnet":"192.168.0.0/24", "netname":"private", "hostname":"router"}'
  #       }
  #       use_redis => false
  #    }
  # }
  # ----------------------------------
  #
  config_name "weblookup"

  config :fields, :validate => :array, :required => true, :default => ["ClientIP"]
  # {"ip":"8.8.8.8","netname":"Google","subnet":"8.8.8.0\/24","hostname":"google-public-dns-a.google.com"}
  # In the query parameter has the <ip> tag will be replaced by the IP address to lookup, other parameters are optional and according to your lookup service. 
  config :destinations, :validate => :array, :required => false, :default => ["message"]

  # {"ip":"8.8.8.8","netname":"Google","subnet":"8.8.8.0\/24","hostname":"google-public-dns-a.google.com"}
  # In the query parameter has the <ip> tag will be replaced by the IP address to lookup, other parameters are optional and according to your lookup service. 
  config :url, :validate => :string, :required => false, :default => "http://localhost/ripe.php?ip=<item>&TOKEN=token"

  # Optional ruby hash with the key as a string and the value as a string in the form of a JSON object. These key's will not be looked up.
  config :list, :validate => :hash, :required => false

  # Optional Redis IP cache
  config :use_redis, :validate => :boolean, :required => false, :default => false
  config :redis_path, :validate => :string, :required => false
  config :redis_expiry, :validate => :number, :required => false, :default => 604800 

  # Optional simplify the message by moving a field to be the new root of the message
  config :normalize, :validate => :boolean, :required => false, :default => false
  config :newroot, :validate => :string, :required => false
  config :roottostrip, :validate => :string, :required => false

  HTTP_OPTIONS = {
      keep_alive_timeout: 300
  }

public
def register
    if use_redis
        unless redis_path.to_s.strip.empty?
            @red = Redis.new(path: redis_path)
        else
            @red = Redis.new()
        end
    end

    # input fields and destinations
    @is_one_destination=false
    if destinations.size == 1
        @logger.info("one destination found, it is #{destinations[0]}")
    @is_one_destination=true
    else
        if destinations.size != fields.size
            @logger.error("Configuration error, there must be an equal amount of destinations and fields, defaulting to using the field as a root for the new values. e.g. if the lookup is done on the value of [\"ClientIP\"] the destination will be [\"ClientIP\"][\"Key\"]")
            destinations=fields
        end
        # add case destination is empty to put the result in under the same field 
    end

    # http connectionpool
    @uri = Addressable::URI.parse(url)
    @uri.merge!(HTTP_OPTIONS)
    #@http = Net::HTTP.new(uri.host, uri.port, HTTP_OPTIONS)
    @uri.port=80 if (@uri.port.nil? && @uri.scheme=="http")
    @uri.port=443 if (@uri.port.nil? && @uri.scheme=="https")
    # find the key where the value is <item>, otherwise just use the value
    @params = @uri.query_values(Hash)
    @params.each do |key, value|
        if value == "\<item\>" 
            @ip=key
        @params.delete(key)
        logger.info("the ip key in the uri is #{@ip}")
        end
    end
    @connpool = ConnectionPool.new(size: 4, timeout: 180) { 
        Net::HTTP.new(@uri.host, @uri.port)
    }
end # def register

def filter(event)
    if destinations[0] == "srcdst"
        # ... do special sauce
        begin
            src = parse(event.get(fields[0]).to_s)
            dst = parse(event.get(fields[1]).to_s)
            srcdst = { :srcnet => src["netname"], :srchost => src["hostname"], :dstnet => dst["netname"], :dsthost => dst["hostname"] }
            event.set("srcdst", srcdst)
            event.get("[srcdst]").each {|k, v| event.set(k, v) }
            event.remove("[srcdst]")
            @logger.trace("processed: #{event.get(fields[0]).to_s} #{src} #{event.get(fields[1]).to_s} #{dst} #{srcdst}")
        rescue Exception => e
            @logger.error(" caught: #{e.message}")
        end
    else
        fields.each_with_index do |field, index|
        # @logger.info(event.get("["+field+"]"))
            begin
                json = parse(event.get(field).to_s)
                event.set("["+destinations[index]+"]", json)
            rescue Exception => e
                @logger.error(" caught: #{e.message}")
            end 
        end
    end
    if @normalize
        replant(event, @newroot)
    end
    # filter_matched should go in the last line of our successful code
    filter_matched(event)
end # def filter



private
def parse(field)
    x = find(field)
    begin
        json = JSON.parse(x)
    rescue
        json = JSON.parse("{\"ip\": \""+field+"\"}")
    end
    # @logger.info("json parse option for field #{field} / #{json}")
end

def find(item)
    res = "{}"
    # Is item in list? (list is an optional array)
    #unless list.nil? 
        # What if the list exists, but item is not on the list?
    #    return list[item]
    #end
    # Is item in redis?
    unless @red.nil?
        res = @red.get(item)
    unless res.nil?
            return res
        end
    end

    # find the key where the value is <item>, otherwise just use the value
    current_uri = @uri
    current_uri.query_values = @params.merge({@ip => item})
    #logger.info(@uri.to_s)
    @connpool.with do |conn|
        http_response = conn.request_get(current_uri)
    res = http_response.read_body if http_response.is_a?(Net::HTTPSuccess)
    if res.eql? "null"
            res = "{}"
        end
    #logger.info(res.to_s)
        unless @red.nil?
            @red.set(item, res)
            @red.expire(item,redis_expiry)
        end
    end
    return res
end

# for legacy
#def normalize(event)
#    event.set("net", JSON.parse(net))
#    event.get("[records][properties]").each {|k, v| event.set(k, v) }
#    event.remove("[records]")
#    event.remove("[message]")
#    return event
#end

def replant(event, newroot)
    begin
      @logger.debug("event: #{event.get(newroot)}")
      event.get(newroot).each {|k, v| event.set(k, v) }
      event.remove(@roottostrip)
    rescue Exception => e
      @logger.info("event failed")
    end
end

# From https://github.com/angel9484/logstash-filter-lookup
def json_loader(data)
    get_map.merge!(JSON.parse(File.read(data)))
end

def csv_loader(data)
    data = CSV.read(data).inject(Hash.new) do |acc, v|
      acc[v[0]] = v[1]
      acc
    end
    get_map.merge!(data)
end

def yml_loader(data)
    get_map.merge!(YAML.load_file(data))
end

end
